File: fop-core/src/main/java/org/apache/fop/configuration/DefaultConfigurationBuilder.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id: Accessibility.java 1343632 2012-05-29 09:48:03Z vhennebert $ */
package org.apache.fop.configuration;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.DOMException;
import org.w3c.dom.Document;
import org.xml.sax.SAXException;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

public class DefaultConfigurationBuilder {

    private static final Log LOG = LogFactory.getLog(DefaultConfigurationBuilder.class.getName());

    public DefaultConfiguration build(InputStream confStream) throws ConfigurationException {
        try {
            DocumentBuilder builder = DefaultConfiguration.DBF.newDocumentBuilder();
            Document document = builder.parse(confStream);
            return new DefaultConfiguration(document.getDocumentElement());
        } catch (DOMException e) {
            throw new ConfigurationException("xml parse error", e);
        } catch (ParserConfigurationException e) {
            throw new ConfigurationException("xml parse error", e);
        } catch (IOException e) {
            throw new ConfigurationException("xml parse error", e);
        } catch (SAXException e) {
            throw new ConfigurationException("xml parse error", e);
        } finally {
            try {
                confStream.close();
            } catch (IOException e) {
                throw new IllegalStateException(e);
            }
        }
    }

    public DefaultConfiguration buildFromFile(File file) throws ConfigurationException {
        try {
            DocumentBuilder builder = DefaultConfiguration.DBF.newDocumentBuilder();
            Document document = builder.parse(file);
            return new DefaultConfiguration(document.getDocumentElement());
        } catch (DOMException e) {
            throw new ConfigurationException("xml parse error", e);
        } catch (ParserConfigurationException e) {
            throw new ConfigurationException("xml parse error", e);
        } catch (IOException e) {
            throw new ConfigurationException("xml parse error", e);
        } catch (SAXException e) {
            throw new ConfigurationException("xml parse error", e);
        }
    }
}

File: dotnet/src/Fop.Configuration/DefaultConfigurationBuilder.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Xml;

namespace Fop.Configuration;

/// <summary>
/// Builds a <see cref="DefaultConfiguration"/> tree by parsing an XML stream or file.
/// <para>Port of <c>org.apache.fop.configuration.DefaultConfigurationBuilder</c>.</para>
/// <para>
/// The Java original used <c>javax.xml.parsers.DocumentBuilder</c>; this port loads an
/// <see cref="XmlDocument"/> via an <see cref="XmlReader"/> configured to match the Java
/// factory: non-validating, comments ignored, entity references expanded. (commons-logging
/// is dropped; the Java LOG field was unused.)
/// </para>
/// </summary>
public class DefaultConfigurationBuilder
{
    private static XmlReaderSettings CreateReaderSettings() => new()
    {
        // Validating(false): no DTD validation. ProhibitDtd would reject inline DTDs that
        // the Java parser tolerated, so DTDs are parsed but not validated.
        DtdProcessing = DtdProcessing.Parse,
        // setIgnoringComments(true)
        IgnoreComments = true,
        // setExpandEntityReferences(true): the default XmlReader behaviour expands entities.
        // setIgnoringElementContentWhitespace(true) only has effect with a DTD/validation,
        // which is disabled here, so whitespace is preserved (matching the Java factory).
        IgnoreWhitespace = false,
        XmlResolver = null,
    };

    /// <summary>Builds a configuration from the given XML stream. The stream is closed afterwards.</summary>
    /// <exception cref="ConfigurationException">If the XML cannot be parsed.</exception>
    public DefaultConfiguration Build(Stream confStream)
    {
        ArgumentNullException.ThrowIfNull(confStream);
        try
        {
            XmlDocument document = DefaultConfiguration.NewDocument();
            using XmlReader reader = XmlReader.Create(confStream, CreateReaderSettings());
            document.Load(reader);
            return new DefaultConfiguration(RequireDocumentElement(document));
        }
        catch (Exception e) when (e is XmlException or IOException)
        {
            throw new ConfigurationException("xml parse error", e);
        }
        finally
        {
            // Mirrors the Java finally block which closed the stream (rethrowing any
            // close failure as an unchecked exception).
            confStream.Dispose();
        }
    }

    /// <summary>Builds a configuration by reading and parsing the given file.</summary>
    /// <exception cref="ConfigurationException">If the file cannot be read or parsed.</exception>
    public DefaultConfiguration BuildFromFile(FileInfo file)
    {
        ArgumentNullException.ThrowIfNull(file);
        try
        {
            XmlDocument document = DefaultConfiguration.NewDocument();
            using FileStream stream = file.OpenRead();
            using XmlReader reader = XmlReader.Create(stream, CreateReaderSettings());
            document.Load(reader);
            return new DefaultConfiguration(RequireDocumentElement(document));
        }
        catch (Exception e) when (e is XmlException or IOException)
        {
            throw new ConfigurationException("xml parse error", e);
        }
    }

    private static XmlElement RequireDocumentElement(XmlDocument document) =>
        document.DocumentElement
            ?? throw new ConfigurationException("xml parse error: no document element");
}
